magda-authorization-api/src/createOpaRouter.ts

```diff
@@ -0,0 +1,141 @@
+import * as express from "express";
+import { Router } from "express";
+import * as _ from "lodash";
+import Database from "./Database";
+import { User } from "@magda/typescript-common/dist/authorization-api/model";
+import * as request from "request";
+import * as bodyParser from "body-parser";
+
+export interface OpaRouterOptions {
+    opaUrl: string;
+    database: Database;
+    jwtSecret: string;
+}
+
+const opaRoutes = [
+    {
+        path: "/compile",
+        method: "post"
+    },
+    {
+        path: "/compile",
+        method: "get"
+    },
+    {
+        path: "/data/*",
+        method: "get"
+    },
+    {
+        path: "/data/*",
+        method: "post"
+    },
+    {
+        path: "/query",
+        method: "get"
+    },
+    {
+        path: "/query",
+        method: "post"
+    }
+];
+
+/**
+ * 
+ * @param options proxyReq: http.ClientRequest,
+      req: http.IncomingMessage,
+      res: http.ServerResponse,
+      options: Server.ServerOptions
+ */
+
+export default function createOpaRouter(options: OpaRouterOptions): Router {
+    const database = options.database;
+    const router: Router = express.Router();
+    const jwtSecret: string = options.jwtSecret;
+    const opaUrl: string = options.opaUrl;
+
+    router.use(bodyParser.json({ type: "application/json" }));
+
+    async function appendUserInfoToInput(
+        req: express.Request,
+        res: express.Response
+    ) {
+        let userInfo: User;
+
+        try {
+            userInfo = await database.getCurrentUserInfo(req, jwtSecret);
```

Comment:
probably need to log this or something if it isn't already? Also won't this mean that the user just gets told they're not allowed to do something even though they are because the database failed somehow?

Reply:
This interface mainly used for providing some background data (`input.user`) for OPA evaluation (we have to decide a role to make the story unified & simple 😭 ). The user won't directly interact with the OPA decision here, OPA policy (every policy serves as an API) may decide to return more info through API and UI may decide to tell the user something case by case. 
Put as Anonymous users as it's close to the meaning of unable to authenticate the user.  
@AlexGilleran 
Do you think it's sufficient if we log the database error at backend via a simple `console.log`?

Reply:
> Put as Anonymous users as it's close to the meaning of unable to authenticate the user.

I'm more inclined to say that a database failure here should cause whatever it depends on to fail too - given that it's all about security it'd be good to be as careful as possible. e.g. if down the line we built a version that disabled the anonymous user and an attacker got access by overloading the database until they're defaulted to an anonymous user anyway.

Reply:
Yeah~ I probably will just throw an error with 500 status code for database errors~
Will add this to latest PR for issue/2155

Reply:
Updated on PR for issue/2155:
https://github.com/magda-io/magda/pull/2184/commits/baf54711bcb872741003a4d1a4f5a8ac14aa8352